pkl-core/src/main/java/org/pkl/core/runtime/VmImportAnalyzer.java

```diff
@@ -31,105 +34,142 @@
-          ExternalReaderProcessException {
+      throws IOException, SecurityManagerException {
     var imports = new TreeMap<URI, Set<ImportGraph.Import>>();
     var resolvedImports = new TreeMap<URI, URI>();
     for (var moduleUri : moduleUris) {
-      analyzeSingle(moduleUri, context, imports, resolvedImports);
+      var moduleKey = context.getModuleResolver().resolve(moduleUri);
+      var resolvedModuleKey = moduleKey.resolve(context.getSecurityManager());
+      analyzeSingle(moduleUri, resolvedModuleKey, context, imports, resolvedImports);
     }
     return new ImportGraph(imports, resolvedImports);
   }
 
   @TruffleBoundary
   private static void analyzeSingle(
       URI moduleUri,
+      ResolvedModuleKey resolvedModuleKey,
       VmContext context,
       Map<URI, Set<ImportGraph.Import>> imports,
-      Map<URI, URI> resolvedImports)
-      throws IOException,
-          URISyntaxException,
-          SecurityManagerException,
-          ExternalReaderProcessException {
+      Map<URI, URI> resolvedImports) {
     var moduleResolver = context.getModuleResolver();
     var securityManager = context.getSecurityManager();
-    var importsInModule = collectImports(moduleUri, moduleResolver, securityManager);
-
+    var collectedImports = collectImports(resolvedModuleKey, moduleResolver, securityManager);
+    var importsInModule = new TreeSet<Import>();
+    for (var imprt : collectedImports) {
+      importsInModule.add(imprt.toImport());
+    }
     imports.put(moduleUri, importsInModule);
-    resolvedImports.put(
-        moduleUri, moduleResolver.resolve(moduleUri).resolve(securityManager).getUri());
-    for (var imprt : importsInModule) {
-      if (imports.containsKey(imprt.uri())) {
+    resolvedImports.put(moduleUri, resolvedModuleKey.getUri());
+    for (var imprt : collectedImports) {
+      if (imports.containsKey(imprt.moduleUri)) {
         continue;
       }
-      analyzeSingle(imprt.uri(), context, imports, resolvedImports);
+      analyzeSingle(imprt.moduleUri, imprt.resolvedModuleKey, context, imports, resolvedImports);
     }
   }
 
-  private static Set<ImportGraph.Import> collectImports(
-      URI moduleUri, ModuleResolver moduleResolver, SecurityManager securityManager)
-      throws IOException,
-          URISyntaxException,
-          SecurityManagerException,
-          ExternalReaderProcessException {
-    var moduleKey = moduleResolver.resolve(moduleUri);
-    var resolvedModuleKey = moduleKey.resolve(securityManager);
+  private static Set<ImportEntry> collectImports(
+      ResolvedModuleKey resolvedModuleKey,
+      ModuleResolver moduleResolver,
+      SecurityManager securityManager) {
     List<Entry> importsAndReads;
+    var moduleKey = resolvedModuleKey.getOriginal();
     try {
-      importsAndReads = ImportsAndReadsParser.parse(moduleKey, resolvedModuleKey);
+      importsAndReads =
+          ImportsAndReadsParser.parse(resolvedModuleKey.getOriginal(), resolvedModuleKey);
```

Comment:
```suggestion
          ImportsAndReadsParser.parse(moduleKey, resolvedModuleKey);
```
Same for line 92 and 97.